sql: set volatilities of OID and REG* types in castMap

mgartner opened the pull request:
#### sql: move cast telemetry counters out of castInfo

Telemetry counters for casts have been moved to a separate data
structure from `castInfo`. This will allow removal of `castInfo` and
`validCasts` once volatility information has been added to `castMap`.
Moving the counters to `castMap` is not viable because counters exist at
the family-level while `castMap` describes casts at the OID-level.

Release note: None

#### sql: add volatility to cast map

This commit adds the `volatility` field to `castMap`, which takes
precedence over the entries in `castInfo`. This allows for more granular
configuration of casts. For now, none of the volatilities have been set
in `castMap`, but they will be in future commits. Once all cast
volatilities and volatility hints have been set in `castMap`, we will
remove `castInfo`.

Release note: None

#### sql: set volatilities of OID and REG* types in castMap

Casting between string types and REG* types is a stable cast, while
casting between string types and OID is immutable. This is now correctly
reflected in the `volatility` field of `castMap`.

Fixes #74216

Release note (bug fix): A bug has been fixed that incorrectly allowed
creating computed column expressions, expression indexes, and partial
index predicate expressions with mutable casts between string types and
the types REGCLASS, REGNAMESPACE, REGPROC, REGPROCEDURE, REGROLE, and
REGTYPE. Creating such computed columns, expression indexes, and partial
indexes is now prohibited. Any tables with these types of expressions
may be corrupt and should be dropped and recreated.

mgartner committed:
sql: add volatility to cast map

This commit adds the `volatility` field to `castMap`, which takes
precedence over the entries in `castInfo`. This allows for more granular
configuration of casts. For now, none of the volatilities have been set
in `castMap`, but they will be in future commits. Once all cast
volatilities and volatility hints have been set in `castMap`, we will
remove `castInfo`.

Release note: None

mgartner committed:
sql: set volatilities of OID and REG* types in castMap

Casting between string types and REG* types is a stable cast, while
casting between string types and OID is immutable. This is now correctly
reflected in the `volatility` field of `castMap`.

Fixes #74216

Release note (bug fix): A bug has been fixed that incorrectly allowed
creating computed column expressions, expression indexes, and partial
index predicate expressions with mutable casts between string types and
the types REGCLASS, REGNAMESPACE, REGPROC, REGPROCEDURE, REGROLE, and
REGTYPE. Creating such computed columns, expression indexes, and partial
indexes is now prohibited. Any tables with these types of expressions
may be corrupt and should be dropped and recreated.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74286)
<!-- Reviewable:end -->

mgartner commented on pkg/sql/opt/norm/testdata/rules/scalar:
```diff
@@ -222,7 +222,7 @@ project
       ├── a.s:4::VARCHAR(2) [as=s:11, outer=(4), immutable]
       ├── a.i:2::INT2::INT8 [as=i:12, outer=(2), immutable]
       ├── a.s:4::CHAR::VARCHAR [as=s:13, outer=(4), immutable]
-      ├── ARRAY[a.i:2, 2]::OIDVECTOR [as=array:14, outer=(2), stable]
```
This is changing because the cast from INT to OID is now marked as immutable. As far as I can tell, it _is_ immutable in Postgres, but I don't know why it wasn't set as such before. Notice the `ignoreVolatilityCheck` is set [here](https://github.com/RaduBerinde/cockroach/blob/123ae8783d0c63e935380bed946a44cd8d332334/pkg/sql/sem/tree/casts.go#L255) because we were diverging from Postgres's volatility of this cast.

mgartner replied:
Ahh... REG* types are part of the OID family ([which they really shouldn't be](https://github.com/cockroachdb/cockroach/blob/ce2796a987266f56fd1adcc1d9cc9aefe7c92c39/pkg/sql/types/types.proto#L185-L186)) and INT -> REG* types are stable. INT -> OID was previously marked as stable because volatilities were labeled at the family level. Now that volatilities can be labeled at the type OID level, INT->OID can be correctly marked as immutable.

mgartner committed:
sql: move cast telemetry counters out of castInfo

Telemetry counters for casts have been moved to a separate data
structure from `castInfo`. This will allow removal of `castInfo` and
`validCasts` once volatility information has been added to `castMap`.
Moving the counters to `castMap` is not viable because counters exist at
the family-level while `castMap` describes casts at the OID-level.

Release note: None

RaduBerinde reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74286#-:-MsLUFJKC6101Re55_xA:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTR!

bors r+
